### DISCUSSION TITLE
Unexpected `immer` middlerware behavior on `delete` key at root of state

### DISCUSSION THREAD

**Discussion by simlmx** [pmndrs/zustand#2480] [Bug report]
This test cases sums it up:
```typescript
describe('test immer delete root key', () => {
  const store = createStore<State>()(immer(() => ({})))

  // State starts empty.
  expect(store.getState()).toEqual({})

  // Set some field.
  store.setState((state) => {
    state.x = 3
  })
  
  expect(store.getState()).toEqual({ x: 3 })

  // Delete field.
  store.setState((state) => {
    delete state.x
  })
  
  // The field should be gone.
  expect(store.getState()).toEqual({}) // <- THIS FAILS and is still equal to {x: 3}
})
```

The last `expect` fails.

If I understand correctly, what is happening is that the two states are [merged together][1]; when using `immer`, shouldn't we always set `replace=true`, since `immer` takes care of merging the states.

Indeed if I modify the `immer` middleware plugin and hardcode `replace=true`, the problem goes away.

[1]: https://github.com/pmndrs/zustand/blob/e00c69f82b9db65dc13b88563cc2f713f6bb2320/src/vanilla.ts#L79

**Comment by dai-shi** [comment]
> since immer takes care of merging the states.

Yeah, if that's the case, `replace=true` sounds reasonable.
Would you mind opening a PR. A test case would be good to add.

**Comment by dbritto-dev** [comment]
@simlmx ping me if you need some help with the PR
